File: codes/week3_topic1_analytical2d.m
% Week 3 - Topic 1: 2D Steady-State Heat Conduction (Analytical Fourier Series)
% T(x,y) = (4*T1/pi) * sum_{k=0}^{kmax} exp(-(2k+1)*pi*y/L).*sin((2k+1)*pi*x/L)/(2k+1)
% Reproduces the lecture MATLAB demo: temperature imagesc + quiver flux field.

function week3_topic1_analytical2d
    close all; clc
    L = 1; T1 = 1; kmax = 30;
    xvec = 0:0.01:1; yvec = 0:0.01:1;
    [X, Y] = meshgrid(xvec, yvec);

    % Convergence at the center
    fprintf('Convergence at (0.5L, 0.5L):\n');
    for km = [0 5 10 20 30]
        fprintf('  kmax = %3d -> T/T1 = %.8f\n', km, tempsum(0.5, 0.5, km, T1, L));
    end

    Tmat = zeros(size(X));
    for k = 0:kmax
        n = 2*k + 1;
        Tmat = Tmat + (4*T1/pi) * exp(-n*pi*Y/L) .* sin(n*pi*X/L) / n;
    end

    figure(1); % temperature distribution
    imagesc(xvec, yvec, Tmat); colorbar; colormap('copper'); caxis([0 1]);
    xlabel('x/L'); ylabel('y/L'); title('Temperature distribution (T/T_1)');

    figure(2); % heat flux field
    spacing = 0.01;
    [DX, DY] = gradient(Tmat, spacing);
    quiver(X(1:5:end,1:5:end), Y(1:5:end,1:5:end), ...
           -DX(1:5:end,1:5:end), -DY(1:5:end,1:5:end), 'r'); hold on;
    contour(X, Y, Tmat);
    xlabel('x/L'); ylabel('y/L'); title('Flux distribution (q/k)');

    fprintf('T/T1 at (0.25,0.10) = %.6f\n', tempsum(0.25, 0.10, kmax, T1, L));
    fprintf('T/T1 at (0.50,0.25) = %.6f\n', tempsum(0.50, 0.25, kmax, T1, L));
end

function T = tempsum(x, y, kmax, T1, L)
    T = 0;
    for k = 0:kmax
        n = 2*k + 1;
        T = T + exp(-n*pi*y/L) * sin(n*pi*x/L) / n;
    end
    T = 4*T1/pi * T;
end

File: codes/week3_topic1_analytical2d.jl
# Week 3 - Topic 1: 2D Steady-State Heat Conduction (Analytical Fourier Series)
# T(x,y) = (4*T1/pi) * sum_{k=0}^{kmax} exp(-(2k+1)*pi*y/L) * sin((2k+1)*pi*x/L) / (2k+1)
# Run: julia week3_topic1_analytical2d.jl

function temperature(x, y; kmax=30, T1=1.0, L=1.0)
    s = 0.0
    for k in 0:kmax
        n = 2k + 1
        s += exp(-n * pi * y / L) * sin(n * pi * x / L) / n
    end
    return 4T1 / pi * s
end

function heatflux(x, y; kmax=30, T1=1.0, L=1.0, kcond=1.0)
    sx = 0.0; sy = 0.0
    for k in 0:kmax
        n = 2k + 1
        e = exp(-n * pi * y / L)
        sx += (pi / L) * e * cos(n * pi * x / L)
        sy += -(pi / L) * e * sin(n * pi * x / L)
    end
    pref = 4T1 / pi
    return (-kcond * pref * sx, -kcond * pref * sy)
end

println("Convergence at (0.5L, 0.5L):")
for kmax in (0, 5, 10, 20, 30)
    Tc = temperature(0.5, 0.5; kmax=kmax)
    println("  kmax = $kmax  ->  T/T1 = $(round(Tc, digits=8))")
end

println("\nSpot values (kmax = 30):")
println("  T/T1 at (0.25, 0.10) = $(round(temperature(0.25, 0.10), digits=6))")
println("  T/T1 at (0.50, 0.25) = $(round(temperature(0.50, 0.25), digits=6))")

qx, qy = heatflux(0.5, 0.25)
println("  flux/k at (0.50, 0.25): qx = $(round(qx, digits=6)), qy = $(round(qy, digits=6))")

# Full normalized field on a 101x101 grid
xs = range(0.0, 1.0; length=101)
ys = range(0.0, 1.0; length=101)
T = [temperature(x, y) for y in ys, x in xs]
println("\nField summary: T/T1 in [$(round(minimum(T), digits=4)), $(round(maximum(T), digits=4))]")
println("BC check: T(0, 0.5) = $(temperature(0.0, 0.5)), T(0.5, 0) = $(round(temperature(0.5, 0.0), digits=6))")
